docs/terminals.md

```diff
@@ -0,0 +1,141 @@
+
+## Terminal Modes
+
+`runc` supports two distinct methods for passing stdio to the container's primary process:
+
+* pass-through
+* new terminal
+
+### Pass-Through
+In pass-through mode, the `stdio` passed to the `runc` call will be passed as is to the primary container. This means that if you do the following:
+
+```
+$ echo input | runc run some_container > /tmp/log.out 2>& /tmp/log.err
+```
+
+Then the stdout of `some_container` will appear in the file `/tmp/log.out`, the stderr will appear in `/tmp/log.err`, and the process will receive input of `input`. `runc`'s stdout was redirected to `/tmp/log.out`, and `runc`, in turn, passed that handle to the container itself.
+
+To invoke pass-through mode, configure your container's `config.json` with `terminal: false` (which is the default).
+
+### New Terminal
+If you do _not_ want to pass your current stdio through to the container, tell it to create a new terminal.
+
+When you do so, `runc` will do the following:
+
+1. Create a new pseudo-terminal (`pty`).
+2. Pass this pty's slave to the container's primary process as its stdio.
+3. Send the pty's master to a process to read/write stdio for the container's primary process (details below).
+
+To invoke new terminal mode, configure your container's `config.json` with `terminal: true`.
+
+## Which Mode To Use
+`runc` itself runs in two modes:
+
+* foreground
+* detached
+
+You can use either terminal mode with either runc mode. However, there are considerations that may indicate preference for one mode over another.
+
+### Foreground
+When running in foreground, `runc` and the container you invoke are most like running a regular foreground process. This is the scenario most likely to be useful for running in pass-through terminal mode. As the example above showed:
+
+```
+$ echo input | runc run some_container > /tmp/log.out 2>& /tmp/log.err
+```
+
+Nonetheless, you still have the option, when running in foreground runc mode, of using New Terminal mode. However, if the process is short-lived, you will not have much time to retrieve the console's fds from the socket and doing anything with it.
+
+### Detached
+When running in detached runc mode, you can run your terminal either in Pass-Through or New-Terminal mode.
+
+You run in detached mode by doing one of the following:
+
+* `runc -d`
+* `runc create` followed by `runc start`, i.e. container lifecycle management
+
+#### Pass-Through Detached
+When running runc detached with terminal Pass-Through, the stdin/stdout/stderr of the calling process will be passed to `runc` and from there to the container's process. However, since `runc` detaches, there are several side effects to take into account:
+
+* The stdout/stderr of the process can become intermixed with the stdout/stderr of the calling process.
+* Any input into the calling process can cause indeterminate issues with which process actually receives the input.
```

Comment:
These notes appear to only be meaningful for shell users (and quite a few other points also seem quite shell-focused). Really this should just be a caution to not use pass-through detached in a shell -- rather than a warning about pass-through detached in general.

Reply:
Is that true? I could have a process with its own stdout/stderr/stdin that is not a shell, and if it calls runc detached with pass-through, then the intermixing will happen, and anything fed into stdin will be indeterminate. Granted, stdin is used most often in shells, but it is not exclusive. I think the issue you described to me (slack? wherever it was...) actually is a general one.